treewide: Add `meta.pkgConfigModules` to a few packages

Ericson2314 opened the pull request:
###### Description of changes

Picking up where https://github.com/NixOS/nixpkgs/pull/214304 left off.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Ericson2314 committed:
treewide: Add `meta.pkgConfigModules` to a few packages

Picking up where https://github.com/NixOS/nixpkgs/pull/214304 left off.

Ericson2314 commented:
CC @growpotkin

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/adding-pkg-config-metadata-to-nixpkgs-packages/25281/1

figsoda commented on pkgs/applications/graphics/ImageMagick/6.x.nix:
```diff
@@ -33,7 +34,7 @@ let
 in
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: with finalAttrs; {
```
`with finalAttrs;` shouldn't be used with `finalAttrs.finalPackage`. if `finalAttrs` is too long, `self` could be an alternative

Ericson2314 replied:
I was replacing `rec` with `with`, but I can do a second pass and get rid of the `with` I suppose.

figsoda replied:
didn't review the rest of the files but this change lgtm

bew commented on pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix:
```diff
@@ -31,6 +32,8 @@ stdenv.mkDerivation rec {
 
   outputs = [ "out" "dev" ];
 
+  passthru.tests.pkg-config = testers.testMetaPkgConfig finalAttrs.finalPackage;
```
Since this is always the same, wouldn't it be simpler (but potentially world rebuilding) to add it to `stdenv.mkDerivation` ?

Ericson2314 replied:
I rather repeat this boilerplate than make mkDerivation more antimodular. The former will get fixed while the latter will cause problems without enough people noticing or caring.

Ericson2314 replied:
https://github.com/NixOS/nixpkgs/issues/215162 tracking issue for this nuisance.

SuperSandro2000 commented on pkgs/applications/graphics/ImageMagick/6.x.nix:
```diff
@@ -109,10 +110,13 @@ stdenv.mkDerivation rec {
     done
   '';
 
+  passthru.tests.pkg-config = testers.testMetaPkgConfig finalAttrs.finalPackage;
```
Do we really need to copy paste this to every package?

jtojnar replied:
We want ofborg to validate it so the alternative would be adding to `make-derivation.nix`.

SuperSandro2000 replied:
How bad would the impact be when it is unused?
IMO we definitely will forget this in some packages.

Ericson2314 replied:
https://github.com/NixOS/nixpkgs/issues/215162 tracking issue for this nuisance.

SuperSandro2000 commented on pkgs/development/libraries/libappindicator/default.nix:
```diff
@@ -54,11 +55,17 @@ stdenv.mkDerivation {
+    pkgConfigModules = {
+      "2" = [ "appindicator-0.1" ];
+      "3" = [ "appindicator3-0.1" ];
+    }.${gtkVersion};
```
We nowhere assert if gtkVersion is 2 or 3. It could also be 4 and then we run into a syntax error which should never happen.

Ericson2314 replied:
See the `if gtkVersion == "2"` above. That is not correct if there are other options.

Ericson2314 replied:
I fixed that in a second commit.

SuperSandro2000 reviewed (changes requested):
Where is ``validatePkgConfig``?

As already noted: ``with finalAttrs`` has the same problems that ``with lib`` has and currently I have the opinion that we should not start to use it anywhere.

Ericson2314 committed:
libappindicator: Clean up code a bit

- Properly through on unsupported GTK versions.

- Mark mono suport broken rather than just having a comment.

Ericson2314 commented:
`validatePkgConfig` is orthogonal to this; I rather not start adding yet as that will require lots of rebuilds. I rather get all these converted and then do `validatePkgConfig` on `staging`.

(I don't really like extra testing in the main build anyways because it adds latency for no good reason. Embrace parallelism; embrace finer-grained derivations.)

Ericson2314 committed:
bzip2: Add `meta.pkgConfigModules` and test

Ericson2314 committed:
cairo: Add `meta.pkgConfigModules` and test

Ericson2314 committed:
libdbusmenu: Add `meta.pkgConfigModules` and test

Odd that the pkg-config names contain these numbers, but they do.

Ericson2314 committed:
libGL: Add `meta.pkgConfigModules` and test

Ericson2314 committed:
fftw: Add `meta.pkgConfigModules` and test

Add `fftwQuad` to `all-packages.nix` too so that variant is tested.

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Ericson2314 committed:
ncurses: Add `meta.pkgConfigModules` and test

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on pkgs/development/libraries/ncurses/default.nix:
```diff
@@ -8,13 +8,14 @@
   pname = "ncurses" + lib.optionalString (abiVersion == "5") "-abi5-compat";
 
-  src = fetchurl {
+  src = let inherit (finalAttrs) version; in fetchurl {
     url = "https://invisible-island.net/archives/ncurses/ncurses-${version}.tar.gz";
```
```suggestion
  src = fetchurl {
    url = "https://invisible-island.net/archives/ncurses/ncurses-${finalAttrs.version}.tar.gz";
```

I am feeling I get really nitty at this point but I would personally prefer this

SuperSandro2000 commented on pkgs/development/libraries/fftw/default.nix:
```diff
@@ -12,15 +12,16 @@
   version = "3.3.10";
 
-  src = fetchurl {
+  src = let inherit (finalAttrs) version; in fetchurl {
     urls = [
       "https://fftw.org/fftw-${version}.tar.gz"
       "ftp://ftp.fftw.org/pub/fftw/fftw-${version}.tar.gz"
```
```suggestion
  src = fetchurl {
    urls = [
      "https://fftw.org/fftw-${finalAttrs.version}.tar.gz"
      "ftp://ftp.fftw.org/pub/fftw/fftw-${finalAttrs.version}.tar.gz"
```
